### ISSUE TITLE
#[must_use] on traits, e.g. on Future

### ISSUE THREAD

**Issue by emilk** [rust-lang/rust#55506]
Putting `#[must_use]` on a trait is currently a no-op (it doesn't seem to do anything, but it is allowed). See [example on playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=ec0405eb63b8d255f94b20a099ba039d).

One important use-case for this would be to have `#[must_use]` on `std::Future`. When returning `impl Future<…>` or `Box<dyn Future<…>>` from a function, it should lead to a warning if that future is not used, as that is almost certainly a bug.

**Comment by Centril** [comment]
cc @varkor

**Comment by zackmdavis** [comment]
(Also see [previous discussion on #51560](https://github.com/rust-lang/rust/issues/51560#issuecomment-397396890).)

**Comment by F001** [comment]
`Fn` `FnMut` `FnOnce` are good candidates as well.

**Comment by Centril** [comment]
@F001 Can you elaborate? If `Output = ()` then I probably don't want the warning...?

**Comment by F001** [comment]
Well, I think this warning means this:
```rust
let c : impl Fn(i32) = ... ;
c(); // `c` must be used. If `c` is discarded, it probably indicates something wrong.

let r = c(); // I think this attribute does not care `r` is used or not
```

**Comment by Centril** [comment]
@F001 Yeah that makes sense.

**Comment by FrankSD** [comment]
Has  this issue been  taken?

**Comment by varkor** [comment]
@FrankSD: no — you're welcome to take this issue! The high-level instructions are in https://github.com/rust-lang/rust/issues/55506#issuecomment-440381411, but feel free to ask any question you have here.

**Comment by varkor** [comment]
@FrankSD: this appears to have been completed in https://github.com/rust-lang/rust/pull/56677. However, at least one of the traits could do with a more informative `#[must_use]` message as highlighted in https://github.com/rust-lang/rust/pull/56677/files#r241236020 (and perhaps the others could be improved too). So you're welcome to open a PR with that improvement!

---

I'm going to close this issue, though, as I think it's been completed satisfactorily.

**Comment by Nemo157** [comment]
The OP mentions both `impl Future` and `Box<dyn Future>`. It appears the former has been implemented, but the latter doesn't warn still.

```rust
#[must_use]
trait Foo {}

impl Foo for () {}

fn foo() -> impl Foo {
    ()
}

fn bar() -> Box<dyn Foo> {
    Box::new(())
}


fn main() {
    foo();
    bar();
}
```

only gives ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d2883e68590f20cafe6aceb7f0a8ed0e))

```
   Compiling playground v0.0.1 (/playground)
warning: unused implementer of `Foo` that must be used
  --> src/main.rs:16:5
   |
16 |     foo();
   |     ^^^^^^
   |
   = note: #[warn(unused_must_use)] on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.68s
     Running `target/debug/playground`
```

**Comment by varkor** [comment]
This was an oversight on my part; I'll open up a PR soon to fix this.

**Comment by Centril** [comment]
It does apply to `let x: dyn Future`.

As for `Box<dyn Trait>` and `#[must_use]` it doesn't seem trivial to me unless you make it a special case for `Box`. We should think of a general solution for type constructors.

**Comment by varkor** [comment]
This is related to (or is maybe just a subissue of) https://github.com/rust-lang/rust/issues/39524. It seems reasonable to special-case `Box` as a stop-gap and try to fix the whole issue at some point.

**Comment by Centril** [comment]
@varkor If you're going for a hack, an attribute based hack seems better, e.g.

```rust
pub struct Box<#[rustc_propagate_must_use] T: ?Sized>(Unique<T>);
```

**Comment by varkor** [comment]
@Centril: I'm not really going for a hack — `Box` is already a special-cased type and I'm simply going to add explicit handling for it into `unused.rs`. I think we'd want to special case it even if we had general support for unused components of structs, to give better error messages.

**Comment by andll** [comment]
@varkor is this going to propagate `#[must_use]` for `Pin<Box<Future>>`? Or just `Box<Future>`?

**Comment by varkor** [comment]
@andll: I don't think this will propagate for `Pin<Box<Future>>`, but I imagine https://github.com/rust-lang/rust/pull/62262 would do.

**Comment by shepmaster** [comment]
I'd like to bring this back up a bit. 

I've been doing some introductory async work, and I've been bitten by the lack of hints to use a boxed future (`futures::future::BoxFuture<'a, T>` / `Pin<Box<dyn Future<Output = T> + 'a + Send>>`) multiple times. This is exacerbated because we cannot use `async fn` in traits, so boxing is a common workaround — the compiler error messages even suggest using [a crate](https://crates.io/crates/async-trait) that automatically performs this rewriting!

**Comment by Nemo157** [comment]
@shepmaster new issue for that appears to be https://github.com/rust-lang/rust/issues/67387
